publish-service/src/main/java/com/ericsson/eiffel/remrem/publish/controller/ProducerController.java

```diff
@@ -111,8 +111,10 @@ public ResponseEntity send(@ApiParam(value = "message protocol", required = true
                 return new ResponseEntity(e.getMessage(), HttpStatus.NOT_FOUND);
             }
         }
-        SendResult result = messageService.send(body, msgService, userDomain, tag, routingKey);
-        return new ResponseEntity(result, messageService.getHttpStatus());
+        synchronized(this) {
+            SendResult result = messageService.send(body, msgService, userDomain, tag, routingKey);
+            return new ResponseEntity(result, messageService.getHttpStatus());
+        }
```

Comment:
As noted in the original PR, make the send method thread-safe instead of synchronizing all callers.

Reply:
I have removed synchronized block and made each method thread-safe using synchronized keyword.

Reply:
No, synchronizing those methods won't solve the race condition. It just makes it more unlikely. The fundamental problem is that send() doesn't return the HTTP status codes but instead stores it (them) in the statusCodes class attribute. Synchronizing the send() methods just means that those methods will execute in serial, but if you have two calls immediately after each other the status codes from the first call will get overwritten by those of the second call. If the code making the first call doesn't call getHttpStatus() quickly enough it'll still get the status codes from the second call.

If you want MessageServiceRMQImpl to be safe for concurrent calls you need to get rid of the (mutable) class attributes.

Reply:
I have checked the MessageServiceRMQImpl class. There are interdependent classes with  MessageServiceRMQImpl. If we make the attributes in this class immutable, then we need to change the dependant class attributes to immutable for which a lot of code refactoring is needed.

The customer is blocked due to this issue and is asking for the delivery immediately.
By using the synchronized block, we are making all the calls synchronized so that the process executes serially. We have tested the changes by publishing bulk of events and it is working as expected. For now, this will be a temporary fix.
I will write an improvement in publish based on your suggestions.

Reply:
Yes, please look into making a proper fix. This PR decreases the risk significantly but it's not zero. And yes, this will require a redesign of MessageServiceRMQImpl and the interface it implements. As it stands it simply doesn't support concurrent requests.